elastic/kibana
# [SharedUX] Whitelabelling POC

## Summary

This PR is a prototype of how whitelabelling may work if we do it via a config file. How it works:
1. It adds a new `theming.yml` file; this file contains configuration for the things we wish to customize
2. it adds a new `whitelabelling` plugin: the responsibility of this plugin is to load the `theming.yml` file, check license and, if appropriate, sets logo and custom mark in chrome; it also registers itself as the only plugin that is allowed to do that (similar to what a security plugin does) => this is to avoid a 3rd party plugin to try and do the same
3. favIcon and pageTitle are set by core directly; since core has no concept of licensing, it is possible this could be abused, but I believe it is minimal risk 

Two major concerns I have with this approach:
1. Logo and mark will have to look good on a dark background; all dark logos will practically be invisible
2. `EuiLoadingSpinner`: this is used throughout the app in many different places; I have not attempted to change them all. if we wish to change this, what is the alternative?

## How to test?
`yarn start -c config/theming.yml`

On a basic license, you should not be able to see logo nor custom mark in Kibana Chrome. 

Once upgraded, logo & mark should change to:
<img width="300" alt="Screenshot 2022-10-06 at 21 26 07" src="https://user-images.githubusercontent.com/1937956/194401480-5d1e2e92-21f8-47aa-ae14-0473e825e8b7.png">

## All comments and feedback are welcomed
 

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
